ttnn/cpp/ttnn/operations/moreh/moreh_adamw/moreh_adamw.hpp

```diff
@@ -0,0 +1,46 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/decorators.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+#include "ttnn/tensor/tensor.hpp"
+
+namespace ttnn::operations::moreh::moreh_adamw {
+
+// A composite operation is an operation that calls multiple operations in sequence
+// It is written using invoke and can be used to call multiple primitive and/or composite operations
+struct MorehAdamw {
+    // The user will be able to call this method as `Tensor output = ttnn::composite_example(input_tensor)` after the op
+    // is registered
+    static std::vector<Tensor> invoke(
+        const Tensor& param_in,
+        const Tensor& grad,
+        const Tensor& exp_avg_in,
+        const Tensor& exp_avg_sq_in,
+
+        const std::optional<float> lr,
+        const std::optional<float> beta1,
+        const std::optional<float> beta2,
+        const std::optional<float> eps,
+        const std::optional<float> weight_decay,
+        const std::optional<uint32_t> step,
+        const std::optional<bool> amsgrad,
+
+        const std::optional<const Tensor> max_exp_avg_sq_in,
+        const std::optional<const Tensor> param_out,
+        const std::optional<const Tensor> exp_avg_out,
+        const std::optional<const Tensor> exp_avg_sq_out,
+        const std::optional<const Tensor> max_exp_avg_sq_out,
+        const std::optional<ttnn::MemoryConfig>& memory_config,
+        std::optional<const DeviceComputeKernelConfig> compute_kernel_config);
+};
+
+}  // namespace ttnn::operations::moreh::moreh_adamw
+
+namespace ttnn {
+constexpr auto moreh_adamw =
+    ttnn::register_operation<"ttnn::moreh_adamw", operations::moreh::moreh_adamw::MorehAdamw>();
```

Comment:
can you please register using with_auto_launch? else this op will have host code executed in main thread

Reply:
Currently, there is a bug with `register_operation_with_auto_launch_op` if the output multiple tensors: `error: static assertion failed due to requirement 'tt::stl::concepts::always_false_v<ttnn::operations::moreh::moreh_adam::MorehAdamw>': Operation is expecting the operator() method to return either a single Tensor or a tuple of Tensor(s). If the operation returns a vector of Tensors, it must implement create_async_output_tensors.` 

Moreh NPU team created a PR: #12356 to fix the issue but hasn't been merged yet. cc: @hschoi4448 
Can we use `register_operation` for now?

Reply:
I think we need to use `register_operation` for now. This is because using with_auto_launch makes the build impossible.
@BuiChiTrung @ayerofieiev-tt

Reply:
@hschoi4448 @BuiChiTrung , the problem described in that PR is about the case when operation returns a vector<optional<Tensor>>. I don't understand how does it block using with_auto_launch here.
Will try to build locally, but please clarify if you have details.

Reply:
`register_operation_with_auto_launch_op` works normally if the output is a Tensor. 

However, if the output is `vector<Tensor>` or `vector<optional>`, an error occurs, which ask us to implement func: `create_async_output_tensors`. Even after implementing the func, we still got new a build error: 
<img width="1812" alt="image" src="https://github.com/user-attachments/assets/ee531417-919d-448f-ab01-e22a6f2bee32">
(the img is from adam operation - very similar to adamw).

Reply:
My code is updated to use `register_operation_with_auto_launch_op`.